zhaquirks/tuya/builder/__init__.py

```diff
@@ -0,0 +1,493 @@
+from zigpy.quirks.v2 import QuirkBuilder, QuirksV2RegistryEntry
+from zigpy.quirks.v2.homeassistant import EntityPlatform, EntityType
+from zigpy.quirks.v2.homeassistant.binary_sensor import BinarySensorDeviceClass
+from zigpy.quirks.v2.homeassistant.number import NumberDeviceClass
+from zigpy.quirks.v2.homeassistant.sensor import SensorDeviceClass, SensorStateClass
+import zigpy.types as t
+from zigpy.zcl import foundation
+from zigpy.zcl.clusters.measurement import (
+    RelativeHumidity,
+    SoilMoisture,
+    TemperatureMeasurement,
+)
+from zigpy.zcl.clusters.smartenergy import Metering
+
+from zhaquirks.tuya import (
+    TUYA_CLUSTER_ID,
+    PowerConfiguration,
+    TuyaLocalCluster,
+    TuyaPowerConfigurationCluster2AAA,
+)
+from zhaquirks.tuya.mcu import DPToAttributeMapping, TuyaMCUCluster, TuyaOnOffNM
+
+
+class TuyaRelativeHumidity(RelativeHumidity, TuyaLocalCluster):
+    """Tuya local RelativeHumidity cluster."""
+
+
+class TuyaTemperatureMeasurement(TemperatureMeasurement, TuyaLocalCluster):
+    """Tuya local TemperatureMeasurement cluster."""
+
+
+class TuyaSoilMoisture(SoilMoisture, TuyaLocalCluster):
+    """Tuya local SoilMoisture cluster with a device RH_MULTIPLIER factor if required."""
+
+
+class TuyaValveWaterConsumed(Metering, TuyaLocalCluster):
+    """Tuya Valve Water consumed cluster."""
+
+    VOLUME_LITERS = 0x0007
+    WATER_METERING = 0x02
+
+    """Setting unit of measurement."""
+    _CONSTANT_ATTRIBUTES = {
+        Metering.AttributeDefs.unit_of_measure.id: VOLUME_LITERS,
+        Metering.AttributeDefs.metering_device_type.id: WATER_METERING,
+    }
+
+    def __init__(self, *args, **kwargs):
+        """Init a TuyaValveWaterConsumed cluster."""
+        super().__init__(*args, **kwargs)
+        self.add_unsupported_attribute(Metering.AttributeDefs.instantaneous_demand.id)
+
+
+class TuyaQuirkBuilder(QuirkBuilder):
+    """Tuya QuirkBuilder."""
+
+    def __init__(
+        self, manufacturer: str, model: str, registry: DeviceRegistry = _DEVICE_REGISTRY
+    ) -> None:
+        """Init the TuyaQuirkBuilder."""
+        self.tuya_data_point_handlers: dict[int, str] = {}
+        self.tuya_dp_to_attribute: dict[int, DPToAttributeMapping] = {}
+        self.new_attributes: dict[int, foundation.ZCLAttributeDef] = {}
+        super().__init__(manufacturer, model, registry)
+
+    def tuya_battery(
+        self,
+        dp_id: int,
+        power_cfg: PowerConfiguration = TuyaPowerConfigurationCluster2AAA,
+        scale: float = 2,
+    ) -> QuirkBuilder:
+        """Add a Tuya Battery Power Configuration."""
+        self.tuya_dp(
+            dp_id,
+            power_cfg.ep_attribute,
+            "battery_percentage_remaining",
+            converter=lambda x: x * scale,
+        )
+        self.adds(power_cfg)
+        return self
+
+    def tuya_metering(
+        self,
+        dp_id: int,
+        metering_cfg: TuyaLocalCluster = TuyaValveWaterConsumed,
```

Comment:
The `Metering` cluster is most often used for tracking energy consumption in devices/quirks.
Most Tuya devices seem to support the ZCL cluster for that at least (just with a manual/different multiplier/divisor), so they do not send it via Tuya datapoints.
I thought about renaming this to `tuya_water_metering`, but that doesn't make a lot of sense, as we can change the `metering_cfg`. So, this is just a note.. It might not be obvious that `tuya_metering` is used for tracking water consumption by default.